cpp/src/arrow/array/builder_dict.cc

```diff
@@ -202,6 +202,9 @@ class internal::DictionaryMemoTable::DictionaryMemoTableImpl {
 
     template <typename DType, typename ArrayType>
     enable_if_memoize<DType, Status> InsertValues(const DType&, const ArrayType& array) {
+      if (array.null_count() > 0) {
+        return Status::Invalid("Cannot yet insert dictionary values containing nulls");
+      }
```

Comment:
I can't remember what motivated to me to add this check here. I'll remove the "yet" since it's unclear whether we need/want to change the behavior to support nulls